### PR TITLE
fix(material/list): fix tabindex="-1" not being maintained when disabled

### PR DESCRIPTION
**Pull request by zarend**
When entire list is disabled, avoid adding list options to the tab order in the following situations:
 - focusin event is triggered on an option. This can happen when the end user clicks on a list option.
 - ngOnInit happens when list is disabled
 - updating selection state

When list is disabled tabindex of every option should always be -1. Tabindex was set to 0 mainly because `_resetActionOption` only considered when list options where disabled, but didn't consider when the list itself is disabled.

Fix issues by setting tabindex to -1 when disabled and making `_handleFocusin` a noop when disabled.

### CODE REVIEW

File: src/material/list/selection-list.ts [hunk 1]
```diff
@@ -373,6 +373,11 @@ export class MatSelectionList
 
   /** Handles focusin events within the list. */
   private _handleFocusin = (event: FocusEvent) => {
+    if (this.disabled) {
```
Comment (crisbeto) [inline src/material/list/selection-list.ts:376]: Maybe I'm missing something, but isn't it too late to reset the tabindex by the time `focusin` fires? It means that focus already entered the list.

Comment (zarend) [reply src/material/list/selection-list.ts:376]: Yes and no. Here's the behavior I observed with a disabled selection list (`<mat-selection-list disabled ...`).

1.  Click on an option
2. (Option because the `document.activeElement` and browser emits focusin event with target of the option under the cursor. Option still has `tabindex="-1")
3. (activeIndex is the index of event target `const activeIndex = this._items.toArray().findIndex(item => item._elementRef.nativeElement.contains(event.target as HTMLElement));`)
4. SelectList focuses the event target via `this._setActiveOption(activeIndex);`
5. (browser emits click event on option.)

So yes, the browser has already focused the option. I even tried adding the `disabled` attribute to the list option, but the browser still focuses on it. It wouldn't say it's too late because the option is focused when browser emits `focusin` event, but it still has `tabindex="-1"` at that time. I don't think we can prevent the browser from focusing it, but we can prevent the selection list from giving it `tabindex="0"`.

Is there any requirement that we have to prevent the browser from focusing a disabled option at all, or is it sufficient to maintain tabindex="-1" in that situation?

Comment (zarend) [reply src/material/list/selection-list.ts:376]: unless maybe we give it `pointer-events: none`. Have this situation come up before? If it has, we can handle it the same way we do in other situations.

Comment (crisbeto) [reply src/material/list/selection-list.ts:376]: Okay, I tried running it and I see the issue that you're solving. I still don't see what how the `focusin` helps in this case. I think that your change to `_resetActiveOption` should be sufficient.

Comment (zarend) [reply src/material/list/selection-list.ts:376]: Okay, I'm willing to take out the changes to `focusin` to get this PR landed.

I debugged it with only the change to `_resetActiveOption` and here's what I noticed. Focusin event would cause tabindex to be 0 on target option. With means you can press up and down arrows to move focus around in the disabled list. Here's a screen capture of that behavior:

https://user-images.githubusercontent.com/7720245/198325569-497e124b-c84f-494a-8997-eb6c1e534bf0.mov

This is the branch and screen capture of what I found while debugging:
https://github.com/angular/components/pull/25882

Since you last reviewed, I changed the skip predicate to always skill when the list itself is disabled. We recently did a similar fix for mat-chip-listbox.

Would you be okay with also having the changes here to address `focusin`? I added a unit test, so we can avoid regressions since it's kinda an edge case.

Comment (crisbeto) [reply src/material/list/selection-list.ts:376]: I think that the correct fix would be not to change the `tabindex` to 0 if the item is disabled. By the time it receives focus it's already too late since the user has landed on it.

Comment (zarend) [reply src/material/list/selection-list.ts:376]: Good catch, I removed the unnecessary call to `_resetActiveOption` inside of `_handleFocusin`. All `_handleFocusin` needs to do is avoid changing the tabindex when the list is disabled. The unit test validates that this implementation is correct.

Does this solution look good to you now?

**Review by crisbeto — APPROVED** [review]
LGTM
